# 🚗 US Accidents Dataset — Data Cleaning & Feature Engineering
---
**Dataset:** US-Accidents: A Countrywide Traffic Accident Dataset  
**Author:** Sobhan Moosavi (Ohio State University)  
**Version:** V8 (2023) &nbsp;|&nbsp; **Records:** ~7.7 Million &nbsp;|&nbsp; **Period:** Feb 2016 – Mar 2023  
**License:** CC BY-NC-SA 4.0 (Research use only)

> **Notebook Goal:** Clean, validate, and feature-engineer the US Accidents dataset to make it analysis and ML-ready.

---


## 📦 1. Imports & Setup

In [1]:
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')


In [3]:
df = pd.read_csv('US_Accidents.csv')

In [4]:
df2 = df.copy()

## 🔍 2. Dataset Overview
Getting an initial look at the data — shape, columns, sample rows, data types, missing values, and duplicates.


### 2.1 Head & Tail

In [5]:
df.head()

,ID,Source,Severity,Start_Time,End_Time,Start_Lat,Start_Lng,End_Lat,End_Lng,Distance(mi),...,Roundabout,Station,Stop,Traffic_Calming,Traffic_Signal,Turning_Loop,Sunrise_Sunset,Civil_Twilight,Nautical_Twilight,Astronomical_Twilight
0,A-1,Source2,3,2016-02-08 05:46:00,2016-02-08 11:00:00,39.865147,-84.058723,NaN,NaN,0.01,...,False,False,False,False,False,False,Night,Night,Night,Night
1,A-2,Source2,2,2016-02-08 06:07:59,2016-02-08 06:37:59,39.928059,-82.831184,NaN,NaN,0.01,...,False,False,False,False,False,False,Night,Night,Night,Day
2,A-3,Source2,2,2016-02-08 06:49:27,2016-02-08 07:19:27,39.063148,-84.032608,NaN,NaN,0.01,...,False,False,False,False,True,False,Night,Night,Day,Day
3,A-4,Source2,3,2016-02-08 07:23:34,2016-02-08 07:53:34,39.747753,-84.205582,NaN,NaN,0.01,...,False,False,False,False,False,False,Night,Day,Day,Day
4,A-5,Source2,2,2016-02-08 07:39:07,2016-02-08 08:09:07,39.627781,-84.188354,NaN,NaN,0.01,...,False,False,False,False,True,False,Day,Day,Day,Day


In [6]:
df.tail()

,ID,Source,Severity,Start_Time,End_Time,Start_Lat,Start_Lng,End_Lat,End_Lng,Distance(mi),...,Roundabout,Station,Stop,Traffic_Calming,Traffic_Signal,Turning_Loop,Sunrise_Sunset,Civil_Twilight,Nautical_Twilight,Astronomical_Twilight
7728389,A-7777757,Source1,2,2019-08-23 18:03:25,2019-08-23 18:32:01,34.00248,-117.37936,33.99888,-117.37094,0.543,...,False,False,False,False,False,False,Day,Day,Day,Day
7728390,A-7777758,Source1,2,2019-08-23 19:11:30,2019-08-23 19:38:23,32.76696,-117.14806,32.76555,-117.15363,0.338,...,False,False,False,False,False,False,Day,Day,Day,Day
7728391,A-7777759,Source1,2,2019-08-23 19:00:21,2019-08-23 19:28:49,33.77545,-117.84779,33.77740,-117.85727,0.561,...,False,False,False,False,False,False,Day,Day,Day,Day
7728392,A-7777760,Source1,2,2019-08-23 19:00:21,2019-08-23 19:29:42,33.99246,-118.40302,33.98311,-118.39565,0.772,...,False,False,False,False,False,False,Day,Day,Day,Day
7728393,A-7777761,Source1,2,2019-08-23 18:52:06,2019-08-23 19:21:31,34.13393,-117.23092,34.13736,-117.23934,0.537,...,False,False,False,False,False,False,Day,Day,Day,Day


### 2.2 Shape & Columns

In [7]:
df.shape

(7728394, 46)

In [8]:
df.columns

Index(['ID', 'Source', 'Severity', 'Start_Time', 'End_Time', 'Start_Lat',
       'Start_Lng', 'End_Lat', 'End_Lng', 'Distance(mi)', 'Description',
       'Street', 'City', 'County', 'State', 'Zipcode', 'Country', 'Timezone',
       'Airport_Code', 'Weather_Timestamp', 'Temperature(F)', 'Wind_Chill(F)',
       'Humidity(%)', 'Pressure(in)', 'Visibility(mi)', 'Wind_Direction',
       'Wind_Speed(mph)', 'Precipitation(in)', 'Weather_Condition', 'Amenity',
       'Bump', 'Crossing', 'Give_Way', 'Junction', 'No_Exit', 'Railway',
       'Roundabout', 'Station', 'Stop', 'Traffic_Calming', 'Traffic_Signal',
       'Turning_Loop', 'Sunrise_Sunset', 'Civil_Twilight', 'Nautical_Twilight',
       'Astronomical_Twilight'],
      dtype='object')

### 2.3 Random Sample

In [9]:
df.sample(10)

,ID,Source,Severity,Start_Time,End_Time,Start_Lat,Start_Lng,End_Lat,End_Lng,Distance(mi),...,Roundabout,Station,Stop,Traffic_Calming,Traffic_Signal,Turning_Loop,Sunrise_Sunset,Civil_Twilight,Nautical_Twilight,Astronomical_Twilight
4555289,A-4589845,Source1,2,2022-12-15 10:29:00,2022-12-15 11:12:00,32.072275,-81.153390,32.072297,-81.141721,0.683,...,False,False,False,False,False,False,Day,Day,Day,Day
3727428,A-3755338,Source1,2,2022-10-12 17:20:55,2022-10-12 18:37:22,34.433868,-82.303115,34.434714,-82.306411,0.197,...,False,False,False,False,False,False,Day,Day,Day,Day
1932745,A-1942608,Source2,3,2019-08-30 19:06:50,2019-08-30 19:35:34,27.878349,-82.657631,NaN,NaN,1.610,...,False,False,False,False,False,False,Day,Day,Day,Day
353045,A-353055,Source2,2,2017-03-31 14:19:17,2017-03-31 15:03:48,33.957504,-117.307388,NaN,NaN,0.010,...,False,False,False,False,False,False,Day,Day,Day,Day
2175278,A-2185144,Source2,2,2019-04-24 06:05:44,2019-04-24 07:54:49,37.498348,-77.405670,NaN,NaN,0.000,...,False,False,True,False,False,False,Night,Day,Day,Day
788763,A-798471,Source2,3,2021-11-30 19:49:18,2021-11-30 20:18:53,34.068218,-118.038551,NaN,NaN,0.000,...,False,False,False,False,False,False,Night,Night,Night,Night
5712104,A-5755038,Source1,2,2021-03-10 06:37:30,2021-03-10 08:01:00,34.053718,-118.236583,34.051876,-118.224221,0.719,...,False,False,False,False,False,False,Day,Day,Day,Day
6908180,A-6957454,Source1,2,2020-10-26 05:05:00,2020-10-26 08:06:40,38.904394,-120.838462,38.905352,-120.839942,0.104,...,False,False,False,False,False,False,Night,Night,Night,Night
1669382,A-1679236,Source2,2,2019-12-26 07:08:11,2019-12-26 08:27:47,35.290218,-80.962921,NaN,NaN,0.000,...,False,False,False,False,False,False,Night,Day,Day,Day
6831077,A-6880254,Source1,2,2020-12-28 21:16:25.000000000,2020-12-28 22:55:34.000000000,34.666675,-83.068905,34.666091,-83.070512,0.100,...,False,False,False,False,False,False,Night,Night,Night,Night


### 2.4 Data Types & Info

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7728394 entries, 0 to 7728393
Data columns (total 46 columns):
 #   Column                 Dtype  
---  ------                 -----  
 0   ID                     object 
 1   Source                 object 
 2   Severity               int64  
 3   Start_Time             object 
 4   End_Time               object 
 5   Start_Lat              float64
 6   Start_Lng              float64
 7   End_Lat                float64
 8   End_Lng                float64
 9   Distance(mi)           float64
 10  Description            object 
 11  Street                 object 
 12  City                   object 
 13  County                 object 
 14  State                  object 
 15  Zipcode                object 
 16  Country                object 
 17  Timezone               object 
 18  Airport_Code           object 
 19  Weather_Timestamp      object 
 20  Temperature(F)         float64
 21  Wind_Chill(F)          float64
 22  Humidity(%)       

### 2.5 Statistical Summary

In [11]:
df.describe()

,Severity,Start_Lat,Start_Lng,End_Lat,End_Lng,Distance(mi),Temperature(F),Wind_Chill(F),Humidity(%),Pressure(in),Visibility(mi),Wind_Speed(mph),Precipitation(in)
count,7.728394e+06,7.728394e+06,7.728394e+06,4.325632e+06,4.325632e+06,7.728394e+06,7.564541e+06,5.729375e+06,7.554250e+06,7.587715e+06,7.551296e+06,7.157161e+06,5.524808e+06
mean,2.212384e+00,3.620119e+01,-9.470255e+01,3.626183e+01,-9.572557e+01,5.618423e-01,6.166329e+01,5.825105e+01,6.483104e+01,2.953899e+01,9.090376e+00,7.685490e+00,8.407210e-03
std,4.875313e-01,5.076079e+00,1.739176e+01,5.272905e+00,1.810793e+01,1.776811e+00,1.901365e+01,2.238983e+01,2.282097e+01,1.006190e+00,2.688316e+00,5.424983e+00,1.102246e-01
min,1.000000e+00,2.455480e+01,-1.246238e+02,2.456601e+01,-1.245457e+02,0.000000e+00,-8.900000e+01,-8.900000e+01,1.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00
25%,2.000000e+00,3.339963e+01,-1.172194e+02,3.346207e+01,-1.177543e+02,0.000000e+00,4.900000e+01,4.300000e+01,4.800000e+01,2.937000e+01,1.000000e+01,4.600000e+00,0.000000e+00
50%,2.000000e+00,3.582397e+01,-8.776662e+01,3.618349e+01,-8.802789e+01,3.000000e-02,6.400000e+01,6.200000e+01,6.700000e+01,2.986000e+01,1.000000e+01,7.000000e+00,0.000000e+00
75%,2.000000e+00,4.008496e+01,-8.035368e+01,4.017892e+01,-8.024709e+01,4.640000e-01,7.600000e+01,7.500000e+01,8.400000e+01,3.003000e+01,1.000000e+01,1.040000e+01,0.000000e+00
max,4.000000e+00,4.900220e+01,-6.711317e+01,4.907500e+01,-6.710924e+01,4.417500e+02,2.070000e+02,2.070000e+02,1.000000e+02,5.863000e+01,1.400000e+02,1.087000e+03,3.647000e+01


### 2.6 Missing Values

In [12]:
df.isnull().sum()

ID                             0
Source                         0
Severity                       0
Start_Time                     0
End_Time                       0
Start_Lat                      0
Start_Lng                      0
End_Lat                  3402762
End_Lng                  3402762
Distance(mi)                   0
Description                    5
Street                     10869
City                         253
County                         0
State                          0
Zipcode                     1915
Country                        0
Timezone                    7808
Airport_Code               22635
Weather_Timestamp         120228
Temperature(F)            163853
Wind_Chill(F)            1999019
Humidity(%)               174144
Pressure(in)              140679
Visibility(mi)            177098
Wind_Direction            175206
Wind_Speed(mph)           571233
Precipitation(in)        2203586
Weather_Condition         173459
Amenity                        0
Bump      

### 2.7 Duplicate Rows

In [13]:
df.duplicated().sum()

np.int64(0)

## 📋 3. Dataset Summary & Column Descriptions

### 3.1 Dataset Overview

| Field | Details |
|---|---|
| **Official Name** | US-Accidents: A Countrywide Traffic Accident Dataset |
| **Author** | Sobhan Moosavi, Ohio State University |
| **Version** | V8 (latest as of 2023) |
| **Total Records** | ~7.7 million accidents |
| **Time Span** | February 2016 – March 2023 |
| **Geography** | 49 US States |
| **Format** | CSV |
| **License** | CC BY-NC-SA 4.0 (Research only) |


### 3.2 Column Descriptions (All 46 Columns)

| # | Column | Type | Description |
|---|---|---|---|
| 1 | `ID` | str | Unique alphanumeric record identifier (e.g., A-3456789). Primary key — no analytical value on its own. |
| 2 | `Source` | str | API/provider that reported the accident (MapQuest, Bing, or both). Useful for studying data collection bias. |
| 3 | `Severity` | int | Traffic impact score from 1 (minor delay) to 4 (road closure). Primary ML target — heavily skewed toward 2. |
| 4 | `Start_Time` | datetime | Exact timestamp when the accident was first detected. Rich source for time-based features. |
| 5 | `End_Time` | datetime | Timestamp when traffic flow returned to normal. Used to calculate accident duration. |
| 6 | `Start_Lat` | float | Latitude of the accident start point (24°–50° for contiguous US). Core geospatial column. |
| 7 | `Start_Lng` | float | Longitude of the accident start point (-125° to -65°). Negative values since US is in western hemisphere. |
| 8 | `End_Lat` | float | Latitude of the far end of the affected road segment. Often missing or equal to Start_Lat for minor accidents. |
| 9 | `End_Lng` | float | Longitude of the far end of the affected road segment. Works with End_Lat to define the impacted road segment. |
| 10 | `Distance(mi)` | float | Length of road (miles) affected by the accident. Proxy for severity — highways tend to have larger values. |
| 11 | `Description` | str | Auto-generated natural language summary from the API. Quality varies by source. Useful for NLP tasks. |
| 12 | `Street` | str | Street or highway name (e.g., I-95 N, Main St). Useful for identifying accident-prone roads. |
| 13 | `City` | str | City where the accident occurred. Miami, Houston, LA, Charlotte are the most frequent. |
| 14 | `County` | str | County of the accident. Useful for regional analysis between city and state granularity. |
| 15 | `State` | str | Two-letter US state abbreviation (e.g., CA, TX). One of the cleanest and most complete columns. |
| 16 | `Zipcode` | str | Postal ZIP code (5-digit or ZIP+4 format). Useful for hyper-local analysis. |
| 17 | `Country` | str | Always `US`. Zero variance — no analytical value. |
| 18 | `Timezone` | str | Timezone of accident location (US/Eastern, US/Central, US/Mountain, US/Pacific). Critical for time analysis. |
| 19 | `Airport_Code` | str | IATA code of nearest weather station (not nearest airport). Links each record to its weather readings. |
| 20 | `Weather_Timestamp` | datetime | Timestamp of the weather observation. May differ from Start_Time — check the gap before using. |
| 21 | `Temperature(F)` | float | Air temperature in Fahrenheit. Ranges ~-30°F to 120°F. Strong correlation with road conditions. |
| 22 | `Wind_Chill(F)` | float | Perceived temperature accounting for wind speed. Only meaningful below 50°F. High null rate in warm records. |
| 23 | `Humidity(%)` | float | Relative humidity (0–100%). High values often correlate with fog or rain. |
| 24 | `Pressure(in)` | float | Atmospheric pressure in inches of mercury (~29.92 inHg normal). Rapid drops indicate incoming storms. |
| 25 | `Visibility(mi)` | float | Horizontal visibility in miles (0 = whiteout, 10 = clear). Below 1 mile strongly linked to severe accidents. |
| 26 | `Wind_Direction` | str | Compass direction wind blows from (N, NE, E, … Calm, Variable). Abbreviated — needs expansion. |
| 27 | `Wind_Speed(mph)` | float | Wind speed in mph. Above 30 mph affects vehicle control. Contains outliers above 300 mph (data errors). |
| 28 | `Precipitation(in)` | float | Precipitation amount in inches. Nulls and zeros often mean the same thing (no precipitation). |
| 29 | `Weather_Condition` | str | Weather label at accident time (Clear, Light Rain, Heavy Snow, Fog, etc.). 100+ unique values — needs grouping. |
| 30 | `Amenity` | bool | POI flag: amenity (park, shop, restaurant) nearby. 76L False vs 1L True — highly imbalanced. |
| 31 | `Bump` | bool | POI flag: speed bump or hump nearby. Only ~3,500 True values out of 7.7M. |
| 32 | `Crossing` | bool | POI flag: pedestrian or road crossing nearby. ~8 lakh True values. |
| 33 | `Give_Way` | bool | POI flag: yield/give-way sign nearby. Only 36,582 True values — highly imbalanced. |
| 34 | `Junction` | bool | POI flag: road junction nearby. 571,342 True — one of the stronger road feature predictors. |
| 35 | `No_Exit` | bool | POI flag: no-exit/dead-end road nearby. Only 19,545 True values. |
| 36 | `Railway` | bool | POI flag: railway level crossing nearby. Only 66,979 True values. |
| 37 | `Roundabout` | bool | POI flag: roundabout nearby. Only 249 True values — most extreme imbalance in dataset. |
| 38 | `Station` | bool | POI flag: transit/bus station nearby. 201,901 True — high pedestrian and congestion zone indicator. |
| 39 | `Stop` | bool | POI flag: stop sign nearby. Useful for identifying stop-sign intersection accidents. |
| 40 | `Traffic_Calming` | bool | POI flag: traffic calming infrastructure nearby (chicanes, raised crosswalks, etc.). Only 7,598 True. |
| 41 | `Traffic_Signal` | bool | POI flag: traffic signal (traffic light) nearby. 1,143,772 True — most analytically important POI column. |
| 42 | `Turning_Loop` | bool | POI flag: turning loop nearby. All False in this dataset — zero variance, drop this column. |
| 43 | `Sunrise_Sunset` | str | Day/Night based on official sunrise/sunset times for that location and date. Clean and complete. |
| 44 | `Civil_Twilight` | str | Day/Night based on civil twilight (sun 0°–6° below horizon). Captures dawn/dusk transition. |
| 45 | `Nautical_Twilight` | str | Day/Night based on nautical twilight (sun 6°–12° below horizon). Darker than civil twilight. |
| 46 | `Astronomical_Twilight` | str | Day/Night based on astronomical twilight (sun 12°–18° below horizon). Closest to full night classification. |


### 3.3 Data Assessment

#### 🔴 Dirty Data Issues

| Category | Issue |
|---|---|
| `accuracy` | `Start_Time` and `End_Time` need dtype conversion to datetime |
| `completeness` | `End_Lat` has 3,402,762 null values |
| `completeness` | `End_Lng` has 3,402,762 null values |
| `accuracy` | `Distance(mi)` has 3,302,161 values = 0 |
| `accuracy` | Values in `Distance(mi)` are in miles — need to consider unit context |
| `completeness` | `Description` has 5 missing values |
| `completeness` | `Street` has 10,869 missing values |
| `completeness` | `City` has 253 missing values |
| `completeness` | `Zipcode` has 1,915 null values |
| `consistency` | `Zipcode` is object type — needs conversion to numeric |
| `completeness` | `Timezone` has 7,808 null values |
| `completeness` | `Airport_Code` has 22,635 null values |
| `accuracy` | `Weather_Timestamp` needs dtype conversion to datetime |
| `completeness` | `Weather_Timestamp` has 120,228 null values |
| `consistency` | `Temperature(F)` needs to be rounded to full integer |
| `completeness` | `Temperature(F)` has 163,853 null values |
| `completeness` | `Wind_Chill(F)` has 1,999,019 null values |
| `completeness` | `Humidity(%)` has 174,144 null values |
| `consistency` | `Humidity(%)` can be rounded and converted to int |
| `completeness` | `Pressure(in)` has 140,679 null values |
| `accuracy` | `Pressure(in)` can be rounded to int |
| `completeness` | `Visibility(mi)` has 177,098 null values |
| `consistency` | `Visibility(mi)` should be rounded to int |
| `consistency` | `Wind_Direction` uses abbreviations (e.g., E for East) — need full form |
| `completeness` | `Wind_Direction` has 175,206 null values |
| `completeness` | `Wind_Speed(mph)` has 571,233 null values |
| `accuracy` | `Wind_Speed(mph)` has 18 values above 300 mph — physically impossible |
| `accuracy` | 9 lakh rows have zero wind speed |
| `completeness` | `Precipitation(in)` has 2,203,586 missing values |
| `consistency` | `Weather_Condition` has 144 categories — 40 with count < 10 |
| `consistency` | 67 lakh rows across 9 main categories — remaining should be grouped as 'Other' |
| `completeness` | `Sunrise_Sunset` has 23,246 null values |
| `completeness` | `Civil_Twilight` has 23,246 null values |
| `completeness` | `Nautical_Twilight` has 23,246 null values |
| `completeness` | `Astronomical_Twilight` has 23,246 null values |

---

#### 🟡 Messy Data Issues

| Issue | Detail |
|---|---|
| `ID` format | Combination of `A-` prefix and row number — only numeric part is needed |
| `Description` content | Combines road name, timing, and short summary — varies by source |
| `Street`, `City`, `Country` | High cardinality — many unique values (informational) |
| `State` abbreviations | e.g., `CA` instead of `California` — needs full name mapping |
| `Zipcode` format | Either 5-digit or 9-digit with hyphen — needs standardization to 5-digit |
| `Timezone` format | Written as `US/Eastern` etc. — `US/` prefix needs to be stripped |
| `Wind_Direction` case | `Calm` and `CALM` are the same — inconsistent casing |
| `Accident Duration` | Can be derived from `Start_Time` and `End_Time` — not directly stored |
| All POI columns | Highly imbalanced True/False — Amenity (76L False), Roundabout (249 True), Turning_Loop (all False) |


## 🧪 4. Data Assessment (Column-by-Column Analysis)

Each column is assessed for:
- **Completeness** — missing/null values
- **Accuracy** — wrong or impossible values  
- **Consistency** — formatting, encoding, unit issues
- **Validity** — values within expected ranges

---


### 4.0 Assessment Trials & Cross-Column Checks

In [14]:
df.columns

Index(['ID', 'Source', 'Severity', 'Start_Time', 'End_Time', 'Start_Lat',
       'Start_Lng', 'End_Lat', 'End_Lng', 'Distance(mi)', 'Description',
       'Street', 'City', 'County', 'State', 'Zipcode', 'Country', 'Timezone',
       'Airport_Code', 'Weather_Timestamp', 'Temperature(F)', 'Wind_Chill(F)',
       'Humidity(%)', 'Pressure(in)', 'Visibility(mi)', 'Wind_Direction',
       'Wind_Speed(mph)', 'Precipitation(in)', 'Weather_Condition', 'Amenity',
       'Bump', 'Crossing', 'Give_Way', 'Junction', 'No_Exit', 'Railway',
       'Roundabout', 'Station', 'Stop', 'Traffic_Calming', 'Traffic_Signal',
       'Turning_Loop', 'Sunrise_Sunset', 'Civil_Twilight', 'Nautical_Twilight',
       'Astronomical_Twilight'],
      dtype='object')

In [15]:
df['Civil_Twilight'].isna().equals(df['Astronomical_Twilight'].isna())

True

In [16]:
df['Civil_Twilight'].isna().equals(df['Nautical_Twilight'].isna())

True

In [17]:
df['Civil_Twilight'].isna().equals(df['Sunrise_Sunset'].isna())

True

In [18]:
df['Astronomical_Twilight'].isna().equals(df['Nautical_Twilight'].isna())

True

### 4.1 Astronomical Twilight

In [19]:
df['Astronomical_Twilight'].value_counts()

Astronomical_Twilight
Day      6377548
Night    1327600
Name: count, dtype: int64

In [20]:
df['Astronomical_Twilight'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 7728394 entries, 0 to 7728393
Series name: Astronomical_Twilight
Non-Null Count    Dtype 
--------------    ----- 
7705148 non-null  object
dtypes: object(1)
memory usage: 59.0+ MB


In [21]:
df['Astronomical_Twilight'].isna().sum()

np.int64(23246)

In [22]:
df['Astronomical_Twilight'].sample(10)

2752226      Day
6743101      Day
1511065      Day
7560744      Day
1069359      Day
1318801      Day
5277481      Day
54273      Night
421194       Day
2752274      Day
Name: Astronomical_Twilight, dtype: object

### 4.2 Nautical Twilight

In [23]:
df['Nautical_Twilight'].value_counts()

Nautical_Twilight
Day      6076156
Night    1628992
Name: count, dtype: int64

In [24]:
df['Nautical_Twilight'].isna().sum()

np.int64(23246)

In [25]:
df['Nautical_Twilight'].sample(10)

5849327      Day
907809       Day
6838506      Day
590090       Day
1026874      Day
6577027    Night
1042914      Day
1930604    Night
7001876      Day
6769363    Night
Name: Nautical_Twilight, dtype: object

### 4.3 Civil Twilight

In [26]:
df['Civil_Twilight'].isna().sum()

np.int64(23246)

In [27]:
df['Civil_Twilight'].value_counts()

Civil_Twilight
Day      5695619
Night    2009529
Name: count, dtype: int64

In [28]:
df['Civil_Twilight'].sample(10)

151583       Day
1794928      Day
612657       Day
1727043      Day
3379605      Day
3227599      Day
3756251      Day
1986090      Day
1094874      Day
3016064    Night
Name: Civil_Twilight, dtype: object

### 4.4 Sunrise / Sunset

In [29]:
df['Sunrise_Sunset'].value_counts()

Sunrise_Sunset
Day      5334553
Night    2370595
Name: count, dtype: int64

In [30]:
df['Sunrise_Sunset'].isna().sum()

np.int64(23246)

In [31]:
df['Sunrise_Sunset'].sample(10)

2333573    Night
7254368      Day
1927693      Day
4212311      Day
4554892      Day
1731440      Day
2309805      Day
4144901      Day
4465407      Day
5983929      Day
Name: Sunrise_Sunset, dtype: object

### 4.5 Turning Loop

In [32]:
df['Turning_Loop'].value_counts()

Turning_Loop
False    7728394
Name: count, dtype: int64

In [33]:
df['Turning_Loop'].isna().sum()

np.int64(0)

In [34]:
df['Turning_Loop'].sample(10)

4998207    False
6917526    False
4198211    False
7190335    False
2370811    False
2103580    False
2064018    False
275184     False
7552940    False
477771     False
Name: Turning_Loop, dtype: bool

### 4.6 Traffic Signal

In [35]:
df['Traffic_Signal'].value_counts()

Traffic_Signal
False    6584622
True     1143772
Name: count, dtype: int64

In [36]:
df['Traffic_Signal'].isna().sum()

np.int64(0)

In [37]:
df['Traffic_Signal'].sample(10)

399826     False
7352467    False
5238155    False
4248875    False
3402789    False
5634152    False
3562702    False
1997110    False
5141823    False
7307053    False
Name: Traffic_Signal, dtype: bool

### 4.7 Traffic Calming

In [38]:
df['Traffic_Calming'].isna().sum()

np.int64(0)

In [39]:
df['Traffic_Calming'].value_counts()

Traffic_Calming
False    7720796
True        7598
Name: count, dtype: int64

In [40]:
df['Traffic_Calming'].sample(10)

3800403    False
6616371    False
4388334    False
3862028    False
2600152    False
563347     False
1523866    False
4861208    False
2120035    False
77632      False
Name: Traffic_Calming, dtype: bool

### 4.8 Station

In [41]:
df['Station'].isna().sum()

np.int64(0)

In [42]:
df['Station'].value_counts()

Station
False    7526493
True      201901
Name: count, dtype: int64

In [43]:
df['Station'].sample(10)

4606159    False
291398     False
2679183    False
6455174    False
2857456    False
2810334    False
2183507    False
283890     False
375987     False
7491884    False
Name: Station, dtype: bool

### 4.9 Roundabout

In [44]:
df['Roundabout'].value_counts()

Roundabout
False    7728145
True         249
Name: count, dtype: int64

In [45]:
df['Roundabout'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 7728394 entries, 0 to 7728393
Series name: Roundabout
Non-Null Count    Dtype
--------------    -----
7728394 non-null  bool 
dtypes: bool(1)
memory usage: 7.4 MB


In [46]:
df['Roundabout'].isna().sum()

np.int64(0)

In [47]:
df['Roundabout'].sample(10)

902237     False
1141933    False
856438     False
7333977    False
5769635    False
4642359    False
5124415    False
5155375    False
1103555    False
582976     False
Name: Roundabout, dtype: bool

### 4.10 Railway

In [48]:
df['Railway'].isna().sum()

np.int64(0)

In [49]:
df['Railway'].value_counts()

Railway
False    7661415
True       66979
Name: count, dtype: int64

In [50]:
df['Railway'].sample(10)

2583419     True
3235665    False
7453309    False
6772464    False
5030804    False
6739250    False
7534983    False
484542     False
4565877    False
260312     False
Name: Railway, dtype: bool

### 4.11 No Exit

In [51]:
df['No_Exit'].describe()

count     7728394
unique          2
top         False
freq      7708849
Name: No_Exit, dtype: object

In [52]:
df['No_Exit'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 7728394 entries, 0 to 7728393
Series name: No_Exit
Non-Null Count    Dtype
--------------    -----
7728394 non-null  bool 
dtypes: bool(1)
memory usage: 7.4 MB


In [53]:
df['No_Exit'].value_counts()

No_Exit
False    7708849
True       19545
Name: count, dtype: int64

In [54]:
df['No_Exit'].sample(10)

3651009    False
1825883    False
5459874    False
4988535    False
5529287    False
712031     False
6398147    False
6654782    False
3608904    False
6435034    False
Name: No_Exit, dtype: bool

### 4.12 Junction

In [55]:
df['Junction'].isna().sum()

np.int64(0)

In [56]:
df['Junction'].value_counts()

Junction
False    7157052
True      571342
Name: count, dtype: int64

In [57]:
df['Junction'].sample(10)

5718414    False
4056988    False
4518858    False
3703553    False
4015300    False
4106679    False
4934801    False
1399872    False
4202791    False
5896664    False
Name: Junction, dtype: bool

### 4.13 Give Way

In [58]:
df['Give_Way'].value_counts()

Give_Way
False    7691812
True       36582
Name: count, dtype: int64

In [59]:
df['Give_Way'].sample(10)

3502001    False
4286711    False
7444560    False
2615107    False
1963545    False
753170     False
6276371    False
7292297    False
1589753    False
2569143    False
Name: Give_Way, dtype: bool

### 4.14 Crossing

In [60]:
df['Crossing'].isna().sum()

np.int64(0)

In [61]:
df['Crossing'].value_counts()

Crossing
False    6854631
True      873763
Name: count, dtype: int64

In [62]:
df['Crossing'].sample(10)

5263266     True
5273988    False
3098105    False
2832690    False
3616542    False
1198365     True
7323736    False
278859     False
5612405    False
3523025     True
Name: Crossing, dtype: bool

### 4.15 Bump

In [63]:
df['Bump'].isna().sum()

np.int64(0)

In [64]:
df['Bump'].value_counts()

Bump
False    7724880
True        3514
Name: count, dtype: int64

In [65]:
df['Bump'].sample(10)

996275     False
1092982    False
1355920    False
372054     False
5476454    False
6356637    False
1011134    False
3499388    False
5831588    False
6519723    False
Name: Bump, dtype: bool

### 4.16 Amenity

In [66]:
df['Amenity'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 7728394 entries, 0 to 7728393
Series name: Amenity
Non-Null Count    Dtype
--------------    -----
7728394 non-null  bool 
dtypes: bool(1)
memory usage: 7.4 MB


In [67]:
df['Amenity'].isna().sum()

np.int64(0)

In [68]:
df['Amenity'].value_counts()

Amenity
False    7632060
True       96334
Name: count, dtype: int64

In [69]:
df['Amenity'].sample(10)

3534951    False
1505201    False
7724565    False
7367448    False
2547838    False
5605845    False
6633462    False
3878166    False
1198027    False
4651213    False
Name: Amenity, dtype: bool

### 4.17 Weather Condition

In [ ]:
# temp_df.sort_values('count' , ascending=False).head(50)

NameError: name 'temp_df' is not defined

In [ ]:
# temp_df = df['Weather_Condition'].value_counts().reset_index()

In [71]:
df['Weather_Condition'].value_counts()

Weather_Condition
Fair                         2560802
Mostly Cloudy                1016195
Cloudy                        817082
Clear                         808743
Partly Cloudy                 698972
                              ...   
Heavy Sleet / Windy                1
Sand / Windy                       1
Heavy Rain Shower / Windy          1
Blowing Snow Nearby                1
Drifting Snow                      1
Name: count, Length: 144, dtype: int64

In [72]:
df['Weather_Condition'].sample(10)

3368469          Overcast
4084859            Cloudy
1295945    Cloudy / Windy
6403522     Mostly Cloudy
6095164              Fair
3994316            Cloudy
4251158              Fair
57581               Clear
3337244     Partly Cloudy
6744266              Fair
Name: Weather_Condition, dtype: object

### 4.18 Precipitation (in)

In [73]:
df['Precipitation(in)'].describe()

count    5.524808e+06
mean     8.407210e-03
std      1.102246e-01
min      0.000000e+00
25%      0.000000e+00
50%      0.000000e+00
75%      0.000000e+00
max      3.647000e+01
Name: Precipitation(in), dtype: float64

In [74]:
df['Precipitation(in)'].isna().sum()

np.int64(2203586)

In [75]:
df['Precipitation(in)'].sample(10)

2398484    NaN
6730396    0.0
2636886    NaN
4066571    0.0
1040761    0.0
1109428    0.0
3653021    0.0
1432030    0.0
6547988    0.0
2029841    0.0
Name: Precipitation(in), dtype: float64

### 4.19 Wind Speed (mph)

In [76]:
df[df['Wind_Speed(mph)'] == 0]

,ID,Source,Severity,Start_Time,End_Time,Start_Lat,Start_Lng,End_Lat,End_Lng,Distance(mi),...,Roundabout,Station,Stop,Traffic_Calming,Traffic_Signal,Turning_Loop,Sunrise_Sunset,Civil_Twilight,Nautical_Twilight,Astronomical_Twilight
43488,A-43489,Source2,3,2016-06-22 23:13:09,2016-06-22 23:58:09,34.000114,-117.063087,NaN,NaN,0.000,...,False,False,False,False,False,False,Night,Night,Night,Night
44266,A-44267,Source2,3,2016-06-25 05:47:46,2016-06-25 06:32:46,34.459503,-118.615616,NaN,NaN,0.000,...,False,False,False,False,False,False,Day,Day,Day,Day
44410,A-44411,Source2,2,2016-06-26 05:41:28,2016-06-26 06:28:19,34.433743,-118.385757,NaN,NaN,0.000,...,False,False,False,False,False,False,Night,Day,Day,Day
44568,A-44569,Source2,2,2016-06-27 09:39:22,2016-06-27 10:18:00,34.365108,-118.503716,NaN,NaN,0.000,...,False,False,False,False,False,False,Day,Day,Day,Day
44570,A-44571,Source2,2,2016-06-27 09:38:53,2016-06-27 10:15:00,34.402683,-118.454987,NaN,NaN,0.000,...,False,False,False,False,False,False,Day,Day,Day,Day
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
7728358,A-7777726,Source1,2,2019-08-23 08:24:31,2019-08-23 08:52:30,34.120400,-117.740320,34.120190,-117.73288,0.426,...,False,False,False,False,False,False,Day,Day,Day,Day
7728361,A-7777729,Source1,2,2019-08-23 12:52:31,2019-08-23 13:20:14,34.023790,-118.276390,34.025760,-118.27529,0.150,...,False,False,False,False,False,False,Day,Day,Day,Day
7728364,A-7777732,Source1,2,2019-08-23 13:42:50,2019-08-23 14:10:06,34.070610,-118.263910,34.069740,-118.26155,0.148,...,False,False,False,False,False,False,Day,Day,Day,Day
7728375,A-7777743,Source1,2,2019-08-23 16:28:17,2019-08-23 16:57:04,34.323752,-118.500833,34.340840,-118.52113,1.654,...,False,False,False,False,False,False,Day,Day,Day,Day


In [77]:
df[df['Wind_Speed(mph)'] > 300]

,ID,Source,Severity,Start_Time,End_Time,Start_Lat,Start_Lng,End_Lat,End_Lng,Distance(mi),...,Roundabout,Station,Stop,Traffic_Calming,Traffic_Signal,Turning_Loop,Sunrise_Sunset,Civil_Twilight,Nautical_Twilight,Astronomical_Twilight
114580,A-114587,Source2,3,2016-06-13 18:28:43,2016-06-13 19:13:43,34.303253,-118.479774,NaN,NaN,0.000,...,False,False,False,False,False,False,Day,Day,Day,Day
114588,A-114595,Source2,3,2016-06-13 18:44:57,2016-06-13 19:29:57,34.320553,-118.495262,NaN,NaN,0.000,...,False,False,False,False,False,False,Day,Day,Day,Day
114633,A-114640,Source2,3,2016-06-13 20:06:45,2016-06-13 20:51:45,34.247803,-118.425835,NaN,NaN,0.000,...,False,False,False,False,False,False,Night,Day,Day,Day
114640,A-114647,Source2,2,2016-06-13 20:19:08,2016-06-13 21:04:08,34.266048,-118.467438,NaN,NaN,0.000,...,False,False,False,False,False,False,Night,Day,Day,Day
114647,A-114654,Source2,3,2016-06-13 20:30:47,2016-06-13 21:15:47,34.266048,-118.450768,NaN,NaN,0.000,...,False,False,False,False,False,False,Night,Day,Day,Day
2638811,A-2648683,Source2,2,2018-08-15 12:56:31,2018-08-15 13:26:16,38.559528,-121.373108,NaN,NaN,0.000,...,False,False,False,False,False,False,Day,Day,Day,Day
2638812,A-2648684,Source2,2,2018-08-15 13:03:11,2018-08-15 13:32:56,38.557919,-121.386116,NaN,NaN,0.000,...,False,False,False,False,False,False,Day,Day,Day,Day
3104760,A-3114639,Source2,2,2017-12-07 15:04:53,2017-12-07 15:49:38,34.030327,-117.650681,NaN,NaN,0.000,...,False,False,False,False,False,False,Day,Day,Day,Day
3144938,A-3154817,Source2,3,2017-12-28 08:35:43,2017-12-28 09:05:27,34.279095,-118.469826,NaN,NaN,0.000,...,False,False,False,False,False,False,Day,Day,Day,Day
3144960,A-3154839,Source2,3,2017-12-28 10:41:44,2017-12-28 11:26:26,34.321411,-118.496750,NaN,NaN,0.000,...,False,False,False,False,False,False,Day,Day,Day,Day


In [78]:
df['Wind_Speed(mph)'].max()

1087.0

In [79]:
df['Wind_Speed(mph)'].describe()

count    7.157161e+06
mean     7.685490e+00
std      5.424983e+00
min      0.000000e+00
25%      4.600000e+00
50%      7.000000e+00
75%      1.040000e+01
max      1.087000e+03
Name: Wind_Speed(mph), dtype: float64

In [80]:
df['Wind_Speed(mph)'].sample(5)

1940098     9.0
1206804     0.0
1132063     6.0
899541     17.0
7561259    11.5
Name: Wind_Speed(mph), dtype: float64

In [81]:
df['Wind_Speed(mph)'].isnull().sum()

np.int64(571233)

### 4.20 Wind Direction

In [82]:
df['Wind_Direction'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 7728394 entries, 0 to 7728393
Series name: Wind_Direction
Non-Null Count    Dtype 
--------------    ----- 
7553188 non-null  object
dtypes: object(1)
memory usage: 59.0+ MB


In [83]:
df['Wind_Direction'].isnull().sum()

np.int64(175206)

In [84]:
df['Wind_Direction'].unique()

array(['Calm', 'SW', 'SSW', 'WSW', 'WNW', 'NW', 'West', 'NNW', 'NNE',
       'South', 'North', 'Variable', 'SE', 'SSE', 'ESE', 'East', 'NE',
       'ENE', 'E', 'W', nan, 'S', 'VAR', 'CALM', 'N'], dtype=object)

In [85]:
df['Wind_Direction'].sample(5)

446414          ENE
3387634       South
3167026    Variable
3013778       North
7013623         WSW
Name: Wind_Direction, dtype: object

### 4.21 Visibility (mi)

In [86]:
df['Visibility(mi)'].isnull().sum()

np.int64(177098)

In [87]:
df['Visibility(mi)'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 7728394 entries, 0 to 7728393
Series name: Visibility(mi)
Non-Null Count    Dtype  
--------------    -----  
7551296 non-null  float64
dtypes: float64(1)
memory usage: 59.0 MB


In [88]:
df['Visibility(mi)'].describe()

count    7.551296e+06
mean     9.090376e+00
std      2.688316e+00
min      0.000000e+00
25%      1.000000e+01
50%      1.000000e+01
75%      1.000000e+01
max      1.400000e+02
Name: Visibility(mi), dtype: float64

In [89]:
df['Visibility(mi)'].sample(10)

2092690    10.0
4001202    10.0
5503296    10.0
2440451     8.0
1956140    10.0
7286002    10.0
3495195    10.0
2156240    10.0
100505      5.0
4774559    10.0
Name: Visibility(mi), dtype: float64

### 4.22 Pressure (in)

In [90]:
df['Pressure(in)'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 7728394 entries, 0 to 7728393
Series name: Pressure(in)
Non-Null Count    Dtype  
--------------    -----  
7587715 non-null  float64
dtypes: float64(1)
memory usage: 59.0 MB


In [91]:
df['Pressure(in)'].describe()

count    7.587715e+06
mean     2.953899e+01
std      1.006190e+00
min      0.000000e+00
25%      2.937000e+01
50%      2.986000e+01
75%      3.003000e+01
max      5.863000e+01
Name: Pressure(in), dtype: float64

In [92]:
df['Pressure(in)'].isnull().sum()

np.int64(140679)

In [93]:
df['Pressure(in)'].sample(10)

3100878    30.29
5766207    29.79
1181802    30.07
5651458    30.07
2870307    30.11
1645169    29.47
189718     30.12
6091199    29.82
2902490    29.95
4165877    28.78
Name: Pressure(in), dtype: float64

### 4.23 Humidity (%)

In [94]:
df['Humidity(%)'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 7728394 entries, 0 to 7728393
Series name: Humidity(%)
Non-Null Count    Dtype  
--------------    -----  
7554250 non-null  float64
dtypes: float64(1)
memory usage: 59.0 MB


In [95]:
df['Humidity(%)'].describe()

count    7.554250e+06
mean     6.483104e+01
std      2.282097e+01
min      1.000000e+00
25%      4.800000e+01
50%      6.700000e+01
75%      8.400000e+01
max      1.000000e+02
Name: Humidity(%), dtype: float64

In [96]:
df['Humidity(%)'].isnull().sum()

np.int64(174144)

In [97]:
df['Humidity(%)'].sample(10)

1898566     55.0
5045040     31.0
3566264     81.0
1607801     50.0
492007      44.0
421172      73.0
3226030    100.0
2305354     97.0
5198526     65.0
7353998     78.0
Name: Humidity(%), dtype: float64

### 4.24 Wind Chill (F)

In [98]:
df['Wind_Chill(F)'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 7728394 entries, 0 to 7728393
Series name: Wind_Chill(F)
Non-Null Count    Dtype  
--------------    -----  
5729375 non-null  float64
dtypes: float64(1)
memory usage: 59.0 MB


In [99]:
df['Wind_Chill(F)'].describe()

count    5.729375e+06
mean     5.825105e+01
std      2.238983e+01
min     -8.900000e+01
25%      4.300000e+01
50%      6.200000e+01
75%      7.500000e+01
max      2.070000e+02
Name: Wind_Chill(F), dtype: float64

In [100]:
df['Wind_Chill(F)'].isnull().sum()

np.int64(1999019)

In [101]:
df['Wind_Chill(F)'].sample(10)

6783298    63.0
4755674    77.0
4617375    89.0
3695020    78.0
6314476    72.0
5501902    84.0
4073338    19.0
4790236    50.0
2718869     NaN
1413410    83.0
Name: Wind_Chill(F), dtype: float64

### 4.25 Temperature (F)

In [102]:
df['Temperature(F)'].isnull().sum()

np.int64(163853)

In [103]:
df['Temperature(F)'].describe()

count    7.564541e+06
mean     6.166329e+01
std      1.901365e+01
min     -8.900000e+01
25%      4.900000e+01
50%      6.400000e+01
75%      7.600000e+01
max      2.070000e+02
Name: Temperature(F), dtype: float64

In [104]:
df['Temperature(F)'].sample(10)

5416402    40.0
5516170    84.0
1509825    63.0
5275324    21.0
2778759    84.2
6048467    69.0
5267670    89.0
448673     50.0
3347221    78.1
7105795    78.0
Name: Temperature(F), dtype: float64

### 4.26 Weather Timestamp

In [105]:
df['Weather_Timestamp'].describe()

count                 7608166
unique                 941331
top       2022-03-13 01:53:00
freq                     1311
Name: Weather_Timestamp, dtype: object

In [106]:
df['Weather_Timestamp'].isnull().sum()

np.int64(120228)

In [107]:
df['Weather_Timestamp'].sample(10)

4387702    2022-12-22 18:29:00
7240625    2019-12-26 06:47:00
3380062    2017-08-21 12:56:00
1245454    2020-11-23 14:53:00
3926067    2022-02-16 13:14:00
1157027    2021-02-04 19:52:00
570747     2022-07-05 08:51:00
431316     2017-04-17 06:35:00
4877213    2022-09-26 08:51:00
4565293    2022-04-27 05:53:00
Name: Weather_Timestamp, dtype: object

### 4.27 Airport Code

In [108]:
df['Airport_Code'].value_counts()

Airport_Code
KCQT    118332
KRDU    107267
KMCJ    101786
KBNA     98926
KCLT     97273
         ...  
KWRL         1
KAXO         1
KY22         1
KAVK         1
KICR         1
Name: count, Length: 2045, dtype: int64

In [109]:
df['Airport_Code'].isnull().sum()

np.int64(22635)

In [110]:
df['Airport_Code'].sample(10)

4740070    KSNA
1931820    KAJO
921026     KEWR
2278063    KATL
4450498    KHWO
139424     KOPF
7278674    KNUQ
5951599    KORL
7148247    KATL
1981237    KTIK
Name: Airport_Code, dtype: object

### 4.28 Timezone

In [111]:
df['Timezone'].sample(10)

2589170    US/Eastern
3541365    US/Pacific
100077     US/Pacific
4309306    US/Pacific
2069055    US/Eastern
6641208    US/Pacific
664284     US/Eastern
2663661    US/Central
4113396    US/Eastern
3485872    US/Eastern
Name: Timezone, dtype: object

In [112]:
df['Timezone'].value_counts()

Timezone
US/Eastern     3580167
US/Pacific     2062984
US/Central     1645616
US/Mountain     431819
Name: count, dtype: int64

In [113]:
df['Timezone'].isnull().sum()

np.int64(7808)

### 4.29 Zipcode

In [114]:
df['Zipcode'].value_counts()

Zipcode
91761         11247
91706         10022
92407          8922
92507          8850
33186          8375
              ...  
24614-6416        1
27051-9335        1
48631-9436        1
39667             1
84660-2639        1
Name: count, Length: 825094, dtype: int64

In [115]:
df['Zipcode'].describe()

count     7726479
unique     825094
top         91761
freq        11247
Name: Zipcode, dtype: object

In [116]:
df['Zipcode'].isnull().sum()

np.int64(1915)

In [117]:
df['Zipcode'].sample(10)

5654706         22030
6025656    96044-9714
2601693         60187
2072280    85008-3104
4621600    47834-7684
2680933         77055
4029941    33147-1502
6601706    17820-1047
1934416    48502-1202
3903412    70119-1619
Name: Zipcode, dtype: object

### 4.30 State

In [118]:
df['State'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 7728394 entries, 0 to 7728393
Series name: State
Non-Null Count    Dtype 
--------------    ----- 
7728394 non-null  object
dtypes: object(1)
memory usage: 59.0+ MB


In [119]:
df['State'].value_counts()

State
CA    1741433
FL     880192
TX     582837
SC     382557
NY     347960
NC     338199
VA     303301
PA     296620
MN     192084
OR     179660
AZ     170609
GA     169234
IL     168958
TN     167388
MI     162191
LA     149701
NJ     140719
MD     140417
OH     118115
WA     108221
AL     101044
UT      97079
CO      90885
OK      83647
MO      77323
CT      71005
IN      67224
MA      61996
WI      34688
KY      32254
NE      28870
MT      28496
IA      26307
AR      22780
NV      21665
KS      20992
DC      18630
RI      16971
MS      15181
DE      14097
WV      13793
ID      11376
NM      10325
NH      10213
WY       3757
ND       3487
ME       2698
VT        926
SD        289
Name: count, dtype: int64

In [120]:
df['State'].isnull().sum()

np.int64(0)

In [121]:
df['State'].sample(10)

6522774    AZ
7505679    MD
7256746    UT
4249354    CA
7149175    MI
1551791    WY
4161939    MD
4096833    GA
1068596    CT
7151753    CO
Name: State, dtype: object

### 4.31 County

In [122]:
df['County'].value_counts()

County
Los Angeles    526851
Miami-Dade     251601
Orange         241275
Harris         181196
Dallas         157024
                ...  
Mcminn              1
Glascock            1
Woods               1
Mellette            1
Ness                1
Name: count, Length: 1871, dtype: int64

In [123]:
df['County'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 7728394 entries, 0 to 7728393
Series name: County
Non-Null Count    Dtype 
--------------    ----- 
7728394 non-null  object
dtypes: object(1)
memory usage: 59.0+ MB


In [124]:
df['County'].describe()

count         7728394
unique           1871
top       Los Angeles
freq           526851
Name: County, dtype: object

In [125]:
df['County'].isnull().sum()

np.int64(0)

In [126]:
df['County'].sample(10)

4941485         San Diego
351115        Los Angeles
3073097           Luzerne
3570465            Queens
5561169             Union
7542677             Duval
6917929              Pima
6559316             Adams
2906864       Los Angeles
6777493    San Bernardino
Name: County, dtype: object

### 4.32 City

In [127]:
df['City'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 7728394 entries, 0 to 7728393
Series name: City
Non-Null Count    Dtype 
--------------    ----- 
7728141 non-null  object
dtypes: object(1)
memory usage: 59.0+ MB


In [128]:
df['City'].value_counts()

City
Miami          186917
Houston        169609
Los Angeles    156491
Charlotte      138652
Dallas         130939
                ...  
Saint Croix         1
Masardis            1
Okaton              1
Wasta               1
Adell               1
Name: count, Length: 13678, dtype: int64

In [129]:
df['City'].isnull().sum()

np.int64(253)

In [130]:
df['City'].sample(10)

4838146       Leesburg
6590966        Atlanta
1175741      Lexington
3538919        Oakland
59817          Fontana
2828722         Orange
880571        Columbia
2626640     Carrollton
6969379     Dorchester
2532117    Grand Blanc
Name: City, dtype: object

### 4.33 Street

In [131]:
df['Street'].value_counts().head(30)

Street
I-95 N      78430
I-95 S      74528
I-5 N       71968
I-10 E      55572
I-10 W      53725
I-5 S       51781
I-80 W      39662
I-80 E      36113
I-405 N     32364
I-75 N      28166
I-75 S      27546
I-90 E      26426
I-15 N      24470
I-94 W      24003
I-90 W      23279
US-101 N    20041
I-95        20028
I-94 E      19940
I-70 E      19697
I-15 S      19230
I-40 W      19034
I-84 E      18955
I-35 N      18761
I-405 S     18318
I-40 E      17800
CA-99 N     17651
I-70 W      17402
I-84 W      17311
I-64 E      16611
I-64 W      16072
Name: count, dtype: int64

In [132]:
df['Street'].describe()

count     7717525
unique     336306
top        I-95 N
freq        78430
Name: Street, dtype: object

In [133]:
df['Street'].isnull().sum()

np.int64(10869)

In [134]:
df['Street'].sample(10)

6341387                      US-50
3667908                 Harwood Pl
850120             Hollywood Fwy S
3863973                   GA-402 E
5432269                      I-5 N
2470075         Westmore Meyers Rd
6414358     Philipsburg Bigler Hwy
1757686              Washington St
5921158            E Northwest Hwy
1099080         E Philadelphia Ave
Name: Street, dtype: object

### 4.34 Description

In [135]:
df['Description'].isnull().sum()

np.int64(5)

In [136]:
df['Description'].sample(10).values[0:5]

array(['Slow traffic on County Hwy-527A from Orlando to FL-528 W due to accident.',
       'Incident on TN-385 EB near MM 2 Drive with caution.',
       'Accident on Boggs and Champlin Rd from West Tehama to Rancho Tehama Rd / Park Ter.',
       '2 right lane closed due to accident on I-91 Northbound between Exits 5 6 US-5 State St and Exit 8 CT-80 Middletown Ave.',
       'Accident on Killian Station Ct at Killian Station Dr.'],
      dtype=object)

### 4.35 Distance (mi)

In [137]:
df['Distance(mi)'].describe()

count    7.728394e+06
mean     5.618423e-01
std      1.776811e+00
min      0.000000e+00
25%      0.000000e+00
50%      3.000000e-02
75%      4.640000e-01
max      4.417500e+02
Name: Distance(mi), dtype: float64

In [138]:
df['Distance(mi)'].isnull().sum()

np.int64(0)

In [139]:
df['Distance(mi)'][df['Distance(mi)'] == 0].count()

np.int64(3302161)

In [140]:
df['Distance(mi)'].sample(10)

2135915    1.970
240601     0.010
3823079    0.995
487157     0.000
4952665    0.252
1129489    0.000
5327449    1.113
3975662    0.068
4600432    2.507
5792711    1.917
Name: Distance(mi), dtype: float64

### 4.36 End Longitude

In [141]:
df['End_Lng'].describe()

count    4.325632e+06
mean    -9.572557e+01
std      1.810793e+01
min     -1.245457e+02
25%     -1.177543e+02
50%     -8.802789e+01
75%     -8.024709e+01
max     -6.710924e+01
Name: End_Lng, dtype: float64

In [142]:
df['End_Lng'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 7728394 entries, 0 to 7728393
Series name: End_Lng
Non-Null Count    Dtype  
--------------    -----  
4325632 non-null  float64
dtypes: float64(1)
memory usage: 59.0 MB


In [143]:
(df['End_Lng'].isna() == df['End_Lat'].isna()).all()

np.True_

In [144]:
demo1 = df['End_Lat'][df['End_Lat'].isna()].index

In [145]:
demo1

Index([      0,       1,       2,       3,       4,       5,       6,       7,
             8,       9,
       ...
       3402752, 3402753, 3402754, 3402755, 3402756, 3402757, 3402758, 3402759,
       3402760, 3402761],
      dtype='int64', length=3402762)

In [146]:
demo2 = df['End_Lng'][df['End_Lng'].isna()].index

In [147]:
(demo1 == demo2)

array([ True,  True,  True, ...,  True,  True,  True])

In [148]:
df['End_Lng'].isnull().sum()

np.int64(3402762)

In [149]:
df['End_Lng'].sample(5)

1539131           NaN
3545367   -118.216290
26335             NaN
6608461    -84.570541
3980560   -118.473282
Name: End_Lng, dtype: float64

### 4.37 End Latitude

In [150]:
df['End_Lat'].max() , df['End_Lat'].min()

(49.075, 24.566013)

In [151]:
df['End_Lat'].describe()

count    4.325632e+06
mean     3.626183e+01
std      5.272905e+00
min      2.456601e+01
25%      3.346207e+01
50%      3.618349e+01
75%      4.017892e+01
max      4.907500e+01
Name: End_Lat, dtype: float64

In [152]:
df['End_Lat'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 7728394 entries, 0 to 7728393
Series name: End_Lat
Non-Null Count    Dtype  
--------------    -----  
4325632 non-null  float64
dtypes: float64(1)
memory usage: 59.0 MB


In [153]:
df['End_Lat'].isnull().sum()

np.int64(3402762)

In [154]:
df['End_Lat'].sample(5)

1539308          NaN
2120255          NaN
6696753    25.890307
3668689    33.206998
2818603          NaN
Name: End_Lat, dtype: float64

### 4.38 Start Longitude

In [155]:
df['Start_Lng'].max() , df['Start_Lng'].min()

(-67.113167, -124.623833)

In [156]:
df['Start_Lng'].describe()

count    7.728394e+06
mean    -9.470255e+01
std      1.739176e+01
min     -1.246238e+02
25%     -1.172194e+02
50%     -8.776662e+01
75%     -8.035368e+01
max     -6.711317e+01
Name: Start_Lng, dtype: float64

In [157]:
df['Start_Lng'].isnull().sum()

np.int64(0)

In [158]:
df['Start_Lng'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 7728394 entries, 0 to 7728393
Series name: Start_Lng
Non-Null Count    Dtype  
--------------    -----  
7728394 non-null  float64
dtypes: float64(1)
memory usage: 59.0 MB


In [159]:
df['Start_Lng'].sample(5)

4246620    -78.610593
6342109    -80.079932
7066608   -121.902949
4804583   -121.808300
2432349    -78.741638
Name: Start_Lng, dtype: float64

### 4.39 Start Latitude

In [160]:
df['Start_Lat'].isnull().sum()

np.int64(0)

In [161]:
df['Start_Lat'].max() , df['Start_Lat'].min()

(49.002201, 24.5548)

In [162]:
df['Start_Lat'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 7728394 entries, 0 to 7728393
Series name: Start_Lat
Non-Null Count    Dtype  
--------------    -----  
7728394 non-null  float64
dtypes: float64(1)
memory usage: 59.0 MB


In [163]:
df['Start_Lat'].sample(5)

6641358    41.397396
3439200    38.819130
3095354    30.812113
5278047    36.752551
7008551    34.898290
Name: Start_Lat, dtype: float64

### 4.40 End Time

In [164]:
df['End_Time'].isnull().sum()

np.int64(0)

In [165]:
df['End_Time'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 7728394 entries, 0 to 7728393
Series name: End_Time
Non-Null Count    Dtype 
--------------    ----- 
7728394 non-null  object
dtypes: object(1)
memory usage: 59.0+ MB


In [166]:
df['End_Time'].sample(5)

2348645    2019-01-24 07:22:42
3748162    2022-02-14 12:57:52
799469     2021-11-22 08:17:00
543477     2022-08-04 15:57:35
3024770    2018-02-26 13:30:04
Name: End_Time, dtype: object

### 4.41 Start Time

In [167]:
df['Start_Time'].isnull().sum()

np.int64(0)

In [168]:
df['Start_Time'].describe()

count                 7728394
unique                6131796
top       2021-01-26 16:16:13
freq                      225
Name: Start_Time, dtype: object

In [169]:
df['Start_Time'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 7728394 entries, 0 to 7728393
Series name: Start_Time
Non-Null Count    Dtype 
--------------    ----- 
7728394 non-null  object
dtypes: object(1)
memory usage: 59.0+ MB


In [170]:
df['Start_Time'].sample(5)

1660508              2020-02-27 07:21:01
4236459    2023-01-25 14:31:17.000000000
2405562              2018-12-26 06:46:05
5482087    2021-09-17 23:52:16.000000000
6040840              2021-02-28 04:21:00
Name: Start_Time, dtype: object

### 4.42 Severity

In [171]:
df['Severity'].isnull().sum()

np.int64(0)

In [172]:
df['Severity'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 7728394 entries, 0 to 7728393
Series name: Severity
Non-Null Count    Dtype
--------------    -----
7728394 non-null  int64
dtypes: int64(1)
memory usage: 59.0 MB


In [173]:
df['Severity'].describe()

count    7.728394e+06
mean     2.212384e+00
std      4.875313e-01
min      1.000000e+00
25%      2.000000e+00
50%      2.000000e+00
75%      2.000000e+00
max      4.000000e+00
Name: Severity, dtype: float64

In [174]:
df['Severity'].value_counts()

Severity
2    6156981
3    1299337
4     204710
1      67366
Name: count, dtype: int64

In [175]:
df['Severity'].duplicated().sum()

np.int64(7728390)

In [176]:
df['Severity'].head()

0    3
1    2
2    2
3    3
4    2
Name: Severity, dtype: int64

In [177]:
df['Severity'].tail()

7728389    2
7728390    2
7728391    2
7728392    2
7728393    2
Name: Severity, dtype: int64

In [178]:
df['Severity'].sample(10)

4554493    2
874640     3
7361380    2
3408582    4
4308911    2
2583925    2
6407567    2
2812562    3
6076574    2
4622675    2
Name: Severity, dtype: int64

### 4.43 Source

In [179]:
df['Source'].isnull().sum()

np.int64(0)

In [180]:
df['Source'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 7728394 entries, 0 to 7728393
Series name: Source
Non-Null Count    Dtype 
--------------    ----- 
7728394 non-null  object
dtypes: object(1)
memory usage: 59.0+ MB


In [181]:
df['Source'].value_counts()

Source
Source1    4325632
Source2    3305373
Source3      97389
Name: count, dtype: int64

In [182]:
df['Source'].describe()

count     7728394
unique          3
top       Source1
freq      4325632
Name: Source, dtype: object

In [183]:
df['Source'].head()

0    Source2
1    Source2
2    Source2
3    Source2
4    Source2
Name: Source, dtype: object

In [184]:
df['Source'].tail()

7728389    Source1
7728390    Source1
7728391    Source1
7728392    Source1
7728393    Source1
Name: Source, dtype: object

In [185]:
df['Source'].sample(10)

2593064    Source2
6309566    Source1
3254517    Source2
3663607    Source1
4653340    Source1
2114969    Source2
7391593    Source1
3460159    Source1
6663712    Source1
2809727    Source2
Name: Source, dtype: object

---
## 🧹 5. Data Cleaning

> Working on `df2` (a copy of the original `df`) to preserve raw data.

### Issues Identified & Actions Taken

| Category | Issues Found | Action |
|---|---|---|
| **Accuracy** | Wrong dtypes for datetime columns | Converted to `datetime64` |
| **Completeness** | Nulls in End_Lat, End_Lng (34L rows) | Filled with Start_Lat/Lng |
| **Completeness** | Nulls in weather columns | Filled with median/mode |
| **Consistency** | Wind direction abbreviations | Expanded to full names |
| **Consistency** | State abbreviations | Mapped to full state names |
| **Consistency** | Timezone format `US/Eastern` | Stripped to `Eastern` |
| **Accuracy** | Wind speed > 300 mph (18 rows) | Clipped to 300 |
| **Validity** | Zipcode as object type | Converted to int32 |

---


### 5.1 Handling Completeness — Filling Null Values

#### End_Lat & End_Lng
> **Strategy:** Filled using `Start_Lat` and `Start_Lng`.  
> **Justification:** 29 out of 34 lakh missing rows have `Distance = 0`, meaning start and end point are the same. For remaining rows, no bearing data is available so start coordinates are the best proxy.


In [186]:
# Verify this overlap
missing_end = df2[df2['End_Lat'].isna()]
print(missing_end['Distance(mi)'].value_counts().head())
# Most likely Distance = 0 dominates here

Distance(mi)
0.00    2900689
0.01     249056
0.01      13367
0.02       6011
0.45       1144
Name: count, dtype: int64


In [187]:
df2['End_Lat'].isna().sum()

np.int64(3402762)

In [188]:
df2['End_Lat'] = df2['End_Lat'].fillna(df2['Start_Lat'])

In [189]:
df2['End_Lat'].isnull().sum()

np.int64(0)

In [190]:
df2['End_Lng'] = df2['End_Lng'].fillna(df2['Start_Lng'])

In [191]:
df2['End_Lng'].isnull().sum()

np.int64(0)

#### Description
> **Strategy:** Filled with `'No Information'` — 5 missing values only.


In [192]:
df2['Description'].isna().sum()

np.int64(5)

In [193]:
df2['Description'] = df2['Description'].fillna('No Information')

In [194]:
df2['Description'].isna().sum()

np.int64(0)

#### Street
> **Strategy:** Filled with `'Not Known'` — 10,869 missing values.


In [195]:
df2['Street'].isna().sum()

np.int64(10869)

In [196]:
df2['Street'] = df2['Street'].fillna('Not Known')

In [197]:
df2['Street'].isna().sum()

np.int64(0)

#### City
> **Strategy:** Filled with `'Not Known'` — 253 missing values.


In [198]:
df2['City'].isna().sum()

np.int64(253)

In [199]:
df2['City'] = df2['City'].fillna('Not Known')

In [200]:
df2['City'].isna().sum()

np.int64(0)

#### Zipcode
> **Strategy:** Filled with `'0'` before type conversion — 1,915 null values.


In [201]:
df2['Zipcode']= df2['Zipcode'].fillna('0')

In [202]:
df2['Zipcode'].isnull().sum()

np.int64(0)

#### Timezone
> **Strategy:** Derived from longitude ranges since the US has only 4 major timezones.  
> `lng >= -80` → Eastern, `lng >= -96` → Central, `lng >= -115` → Mountain, else → Pacific


In [203]:
df2['Timezone'].isnull().sum()

np.int64(7808)

In [204]:
## works on small dataset


# def fillTimezone(row):

#     if pd.isna(row['Timezone']):
#         lng = row['Start_Lng']

#         if lng >= -80: return 'US/Eastern'
#         elif lng >= -96: return 'US/Central'
#         elif lng >= -115: return 'US/Mountain'
#         else : return 'US/Pacific'

#     return row['Timezone']

In [205]:
conditions = [
    df2['Start_Lng'] >= -80,
    df2['Start_Lng'] >= -96,
    df2['Start_Lng'] >= -115,
    df2['Start_Lng'] < -115
]

choices = ['US/Eastern', 'US/Central', 'US/Mountain', 'US/Pacific']

df2['Timezone'] = df2['Timezone'].fillna(
    pd.Series(np.select(conditions, choices, default='US/Pacific'), index=df2.index)
)

In [206]:
df2['Timezone'].isnull().sum()

np.int64(0)

#### Airport Code
> **Strategy:** Filled with `'Not Known'` — 22,635 null values.


In [207]:
df2['Airport_Code'].isna().sum()

np.int64(22635)

In [208]:
df2['Airport_Code'] = df2['Airport_Code'].fillna('Not Known')

In [209]:
df2['Airport_Code'].isna().sum()

np.int64(0)

#### Weather Timestamp
> **Strategy:** Filled with `Start_Time` — 120,228 null values. Weather at accident time is the best approximation.


In [210]:
df2['Weather_Timestamp'].isnull().sum()

np.int64(120228)

In [211]:
df2['Weather_Timestamp'] = df2['Weather_Timestamp'].astype('datetime64[ns]')

In [212]:
df2['Weather_Timestamp'] = df2['Weather_Timestamp'].fillna(df2['Start_Time'])

In [213]:
df2['Weather_Timestamp'].isnull().sum()

np.int64(0)

#### Numeric Weather Columns
> **Strategy:** Filled with **median** (not mean) because these columns have outliers and skewed distributions.  
> Columns: `Temperature(F)`, `Wind_Chill(F)`, `Humidity(%)`, `Pressure(in)`, `Visibility(mi)`, `Wind_Speed(mph)`, `Precipitation(in)`


In [214]:
df2['Temperature(F)'].isna().sum()

np.int64(163853)

In [215]:
df2['Temperature(F)'] = df2['Temperature(F)'].fillna(df2['Temperature(F)'].median())

In [216]:
df2['Temperature(F)'].isna().sum()

np.int64(0)

#### Wind_Chill

In [217]:
df2['Wind_Chill(F)'].isna().sum()

np.int64(1999019)

In [218]:
df2['Wind_Chill(F)'] = df2['Wind_Chill(F)'].fillna(df2['Wind_Chill(F)'].median() )

In [219]:
df2['Wind_Chill(F)'].isna().sum()

np.int64(0)

#### 'Humidity(%)' , 'Pressure(in)', 'Visibility(mi)' , 'Wind_Speed(mph)', 'Precipitation(in)' , 'Weather_Condition' , 'Wind_Direction'

In [220]:
Num_cols = ['Humidity(%)' , 'Pressure(in)', 'Visibility(mi)' , 'Wind_Speed(mph)', 'Precipitation(in)' ]
cat_cols = ['Weather_Condition' , 'Wind_Direction']

In [221]:
for col in Num_cols:
    print(col , ':' , df2[col].isnull().sum())

Humidity(%) : 174144
Pressure(in) : 140679
Visibility(mi) : 177098
Wind_Speed(mph) : 571233
Precipitation(in) : 2203586


In [222]:
for col in Num_cols:
    df2[col] = df2[col].fillna(df2[col].median())

In [223]:
for col in Num_cols:
    print(col , ':' , df2[col].isnull().sum())

Humidity(%) : 0
Pressure(in) : 0
Visibility(mi) : 0
Wind_Speed(mph) : 0
Precipitation(in) : 0


In [224]:
for col in cat_cols:
    print(col , ':' , df2[col].isnull().sum())

Weather_Condition : 173459
Wind_Direction : 175206


In [225]:
for col in cat_cols:
    df2[col] = df2[col].fillna(df2[col].mode()[0])

In [226]:
for col in cat_cols:
    print(col , ':' , df2[col].isnull().sum())

Weather_Condition : 0
Wind_Direction : 0


#### 'Sunrise_Sunset', 'Civil_Twilight', 'Nautical_Twilight', 'Astronomical_Twilight'

In [227]:
cols = ['Sunrise_Sunset', 'Civil_Twilight', 'Nautical_Twilight',
       'Astronomical_Twilight']

In [228]:
for col in cols:
    print( col ,':' , df2[col].isna().sum())

Sunrise_Sunset : 23246
Civil_Twilight : 23246
Nautical_Twilight : 23246
Astronomical_Twilight : 23246


In [229]:
for col in cols:
    df2[col] = df2[col].fillna(df2[col].mode()[0])

In [230]:
for col in cols:
    print( col ,':' , df2[col].isna().sum())

Sunrise_Sunset : 0
Civil_Twilight : 0
Nautical_Twilight : 0
Astronomical_Twilight : 0


#### Twilight & Sunrise/Sunset Columns  
> **Strategy:** Filled with **mode** — categorical Day/Night columns, mode is appropriate.


---
### 5.2 Handling Messy Data — Consistency & Formatting Fixes


#### ID Column
> Stripped the `A-` prefix, keeping only the numeric part. The letter prefix adds no analytical value.


In [231]:
df2['ID']

0                A-1
1                A-2
2                A-3
3                A-4
4                A-5
             ...    
7728389    A-7777757
7728390    A-7777758
7728391    A-7777759
7728392    A-7777760
7728393    A-7777761
Name: ID, Length: 7728394, dtype: object

In [232]:
df2['ID'] = df2['ID'].str[2:]


In [233]:
df2['ID']

0                1
1                2
2                3
3                4
4                5
            ...   
7728389    7777757
7728390    7777758
7728391    7777759
7728392    7777760
7728393    7777761
Name: ID, Length: 7728394, dtype: object

#### State Column
> Mapped abbreviations to full state names. e.g., `CA` → `California`, `TX` → `Texas`.


In [234]:
df2['State']

0          OH
1          OH
2          OH
3          OH
4          OH
           ..
7728389    CA
7728390    CA
7728391    CA
7728392    CA
7728393    CA
Name: State, Length: 7728394, dtype: object

In [235]:
us_states = {
    'AL': 'Alabama', 'AK': 'Alaska', 'AZ': 'Arizona', 'AR': 'Arkansas',
    'CA': 'California', 'CO': 'Colorado', 'CT': 'Connecticut', 'DE': 'Delaware',
    'FL': 'Florida', 'GA': 'Georgia', 'HI': 'Hawaii', 'ID': 'Idaho',
    'IL': 'Illinois', 'IN': 'Indiana', 'IA': 'Iowa', 'KS': 'Kansas',
    'KY': 'Kentucky', 'LA': 'Louisiana', 'ME': 'Maine', 'MD': 'Maryland',
    'MA': 'Massachusetts', 'MI': 'Michigan', 'MN': 'Minnesota', 'MS': 'Mississippi',
    'MO': 'Missouri', 'MT': 'Montana', 'NE': 'Nebraska', 'NV': 'Nevada',
    'NH': 'New Hampshire', 'NJ': 'New Jersey', 'NM': 'New Mexico', 'NY': 'New York',
    'NC': 'North Carolina', 'ND': 'North Dakota', 'OH': 'Ohio', 'OK': 'Oklahoma',
    'OR': 'Oregon', 'PA': 'Pennsylvania', 'RI': 'Rhode Island', 'SC': 'South Carolina',
    'SD': 'South Dakota', 'TN': 'Tennessee', 'TX': 'Texas', 'UT': 'Utah',
    'VT': 'Vermont', 'VA': 'Virginia', 'WA': 'Washington', 'WV': 'West Virginia',
    'WI': 'Wisconsin', 'WY': 'Wyoming', 'DC': 'District of Columbia'
}

In [236]:
df2['State'] = df2['State'].map(us_states)

#### Zipcode Column
> Standardized to 5-digit format (trimmed 9-digit hyphenated codes) and converted to `int32`.


In [237]:
df2['Zipcode']

0               45424
1          43068-3402
2               45176
3               45417
4               45459
              ...    
7728389         92501
7728390         92108
7728391         92866
7728392         90230
7728393         92346
Name: Zipcode, Length: 7728394, dtype: object

In [238]:
df2['Zipcode'] = df2['Zipcode'].str[0:5]

In [239]:
df2['Zipcode'] = df2['Zipcode'].astype('int32')

In [240]:
df2['Zipcode']

0          45424
1          43068
2          45176
3          45417
4          45459
           ...  
7728389    92501
7728390    92108
7728391    92866
7728392    90230
7728393    92346
Name: Zipcode, Length: 7728394, dtype: int32

#### Timezone Column
> Removed `US/` prefix. e.g., `US/Eastern` → `Eastern`.


In [241]:
df2['Timezone']

0          US/Eastern
1          US/Eastern
2          US/Eastern
3          US/Eastern
4          US/Eastern
              ...    
7728389    US/Pacific
7728390    US/Pacific
7728391    US/Pacific
7728392    US/Pacific
7728393    US/Pacific
Name: Timezone, Length: 7728394, dtype: object

In [242]:
df2['Timezone'] = df2['Timezone'].str[3:]

In [243]:
df2['Timezone']

0          Eastern
1          Eastern
2          Eastern
3          Eastern
4          Eastern
            ...   
7728389    Pacific
7728390    Pacific
7728391    Pacific
7728392    Pacific
7728393    Pacific
Name: Timezone, Length: 7728394, dtype: object

#### Wind Direction Column
> Standardized case (`Calm` → `CALM`), then expanded abbreviations to full directional names.  
> e.g., `N` → `North`, `SW` → `Southwest`.


In [244]:
df2['Wind_Direction'].value_counts()

Wind_Direction
CALM        1136830
S            419989
SSW          384840
W            383913
WNW          378781
NW           369352
Calm         368557
SW           364470
WSW          353806
SSE          349110
NNW          333427
N            307151
SE           294901
E            278914
ESE          268311
NE           258639
ENE          258474
NNE          255311
VAR          250566
South        177375
West         164751
North        153385
Variable     113996
East         103545
Name: count, dtype: int64

In [245]:
df2['Wind_Direction'] = df2['Wind_Direction'].str.replace('Calm' , 'CALM')

In [246]:
df2['Wind_Direction'].value_counts()

Wind_Direction
CALM        1505387
S            419989
SSW          384840
W            383913
WNW          378781
NW           369352
SW           364470
WSW          353806
SSE          349110
NNW          333427
N            307151
SE           294901
E            278914
ESE          268311
NE           258639
ENE          258474
NNE          255311
VAR          250566
South        177375
West         164751
North        153385
Variable     113996
East         103545
Name: count, dtype: int64

In [247]:
df2['End_Time']

0          2016-02-08 11:00:00
1          2016-02-08 06:37:59
2          2016-02-08 07:19:27
3          2016-02-08 07:53:34
4          2016-02-08 08:09:07
                  ...         
7728389    2019-08-23 18:32:01
7728390    2019-08-23 19:38:23
7728391    2019-08-23 19:28:49
7728392    2019-08-23 19:29:42
7728393    2019-08-23 19:21:31
Name: End_Time, Length: 7728394, dtype: object

In [248]:
df2['Start_Time']

0          2016-02-08 05:46:00
1          2016-02-08 06:07:59
2          2016-02-08 06:49:27
3          2016-02-08 07:23:34
4          2016-02-08 07:39:07
                  ...         
7728389    2019-08-23 18:03:25
7728390    2019-08-23 19:11:30
7728391    2019-08-23 19:00:21
7728392    2019-08-23 19:00:21
7728393    2019-08-23 18:52:06
Name: Start_Time, Length: 7728394, dtype: object

In [249]:
df2['Accident_Time'] = df2['End_Time'].astype('datetime64[ns]') - df2['Start_Time'].astype('datetime64[ns]')

In [250]:
df2['Accident_Time']

0         0 days 05:14:00
1         0 days 00:30:00
2         0 days 00:30:00
3         0 days 00:30:00
4         0 days 00:30:00
                ...      
7728389   0 days 00:28:36
7728390   0 days 00:26:53
7728391   0 days 00:28:28
7728392   0 days 00:29:21
7728393   0 days 00:29:25
Name: Accident_Time, Length: 7728394, dtype: timedelta64[ns]

In [251]:
df2['End_Time'] = df2['End_Time'].astype('datetime64[ns]')
df2['Start_Time'] = df2['Start_Time'].astype('datetime64[ns]')

In [252]:
df2['End_Time']

0         2016-02-08 11:00:00
1         2016-02-08 06:37:59
2         2016-02-08 07:19:27
3         2016-02-08 07:53:34
4         2016-02-08 08:09:07
                  ...        
7728389   2019-08-23 18:32:01
7728390   2019-08-23 19:38:23
7728391   2019-08-23 19:28:49
7728392   2019-08-23 19:29:42
7728393   2019-08-23 19:21:31
Name: End_Time, Length: 7728394, dtype: datetime64[ns]

In [253]:
df['Start_Time']

0          2016-02-08 05:46:00
1          2016-02-08 06:07:59
2          2016-02-08 06:49:27
3          2016-02-08 07:23:34
4          2016-02-08 07:39:07
                  ...         
7728389    2019-08-23 18:03:25
7728390    2019-08-23 19:11:30
7728391    2019-08-23 19:00:21
7728392    2019-08-23 19:00:21
7728393    2019-08-23 18:52:06
Name: Start_Time, Length: 7728394, dtype: object

In [254]:
df.columns

Index(['ID', 'Source', 'Severity', 'Start_Time', 'End_Time', 'Start_Lat',
       'Start_Lng', 'End_Lat', 'End_Lng', 'Distance(mi)', 'Description',
       'Street', 'City', 'County', 'State', 'Zipcode', 'Country', 'Timezone',
       'Airport_Code', 'Weather_Timestamp', 'Temperature(F)', 'Wind_Chill(F)',
       'Humidity(%)', 'Pressure(in)', 'Visibility(mi)', 'Wind_Direction',
       'Wind_Speed(mph)', 'Precipitation(in)', 'Weather_Condition', 'Amenity',
       'Bump', 'Crossing', 'Give_Way', 'Junction', 'No_Exit', 'Railway',
       'Roundabout', 'Station', 'Stop', 'Traffic_Calming', 'Traffic_Signal',
       'Turning_Loop', 'Sunrise_Sunset', 'Civil_Twilight', 'Nautical_Twilight',
       'Astronomical_Twilight'],
      dtype='object')

In [255]:
## Handling Temprature

In [256]:
df2['Temperature(F)'] = round((df2['Temperature(F)']))

In [257]:
df2['Temperature(F)'].describe()

count    7.728394e+06
mean     6.171284e+01
std      1.881424e+01
min     -8.900000e+01
25%      5.000000e+01
50%      6.400000e+01
75%      7.600000e+01
max      2.070000e+02
Name: Temperature(F), dtype: float64

In [258]:
df2['Temperature(F)'] = df2['Temperature(F)'].astype('int32')

In [259]:
## Handling pressure

In [260]:
df2['Pressure(in)']


0          29.68
1          29.65
2          29.67
3          29.64
4          29.65
           ...  
7728389    28.92
7728390    29.39
7728391    29.74
7728392    29.62
7728393    28.63
Name: Pressure(in), Length: 7728394, dtype: float64

In [261]:
df2['Pressure(in)'] = round(df2['Pressure(in)'])

In [262]:
df2['Pressure(in)'] = df2['Pressure(in)'].astype('int16')

In [263]:
## Visisbility 

In [264]:
df2['Visibility(mi)'].describe()

count    7.728394e+06
mean     9.111221e+00
std      2.660819e+00
min      0.000000e+00
25%      1.000000e+01
50%      1.000000e+01
75%      1.000000e+01
max      1.400000e+02
Name: Visibility(mi), dtype: float64

In [265]:
df2['Visibility(mi)'] = round(df2['Visibility(mi)'])

In [266]:
df2['Visibility(mi)'] = df2['Visibility(mi)'].astype('int16')

In [267]:
## wind direction

In [268]:
df2['Wind_Direction'].value_counts()

Wind_Direction
CALM        1505387
S            419989
SSW          384840
W            383913
WNW          378781
NW           369352
SW           364470
WSW          353806
SSE          349110
NNW          333427
N            307151
SE           294901
E            278914
ESE          268311
NE           258639
ENE          258474
NNE          255311
VAR          250566
South        177375
West         164751
North        153385
Variable     113996
East         103545
Name: count, dtype: int64

In [269]:
mapping = {
    'S' : 'South',
    'W' : 'West',
    'E' : 'East',
    'N' : 'North',
    'SSW' : 'South South West',
    'WNW' : 'West North West',
    'NW' :'North West',
    'SW' :'South West',
    'WSW' : 'West South West',
    'SSE': 'South South East',
    'NNW' : 'North North West',
    'SE' : 'South East',
    'ESE' : 'East South East',
    'NE' : 'North East',
    'ENE' : 'East North East',
    'NNE' : 'North North East',
    'CALM' : 'CALM',
    'VAR' : 'Varaible',
    'Varaible' : 'Varaible',
    'East' : 'East',
    'West' : "West",
    'South' : 'South',
    'North' : 'North'
}

In [270]:
df2['Wind_Direction']

0          CALM
1          CALM
2            SW
3            SW
4            SW
           ... 
7728389       W
7728390      SW
7728391     SSW
7728392      SW
7728393      SW
Name: Wind_Direction, Length: 7728394, dtype: object

In [271]:
df2['Wind_Direction'] = df2['Wind_Direction'].map(mapping)

#### Start Time & End Time
> Converted to `datetime64` dtype. Calculated `Accident_Time` = `End_Time - Start_Time`.


In [272]:
df2['End_Time']

0         2016-02-08 11:00:00
1         2016-02-08 06:37:59
2         2016-02-08 07:19:27
3         2016-02-08 07:53:34
4         2016-02-08 08:09:07
                  ...        
7728389   2019-08-23 18:32:01
7728390   2019-08-23 19:38:23
7728391   2019-08-23 19:28:49
7728392   2019-08-23 19:29:42
7728393   2019-08-23 19:21:31
Name: End_Time, Length: 7728394, dtype: datetime64[ns]

In [273]:
df2['Start_Time']

0         2016-02-08 05:46:00
1         2016-02-08 06:07:59
2         2016-02-08 06:49:27
3         2016-02-08 07:23:34
4         2016-02-08 07:39:07
                  ...        
7728389   2019-08-23 18:03:25
7728390   2019-08-23 19:11:30
7728391   2019-08-23 19:00:21
7728392   2019-08-23 19:00:21
7728393   2019-08-23 18:52:06
Name: Start_Time, Length: 7728394, dtype: datetime64[ns]

In [274]:
df2['Accident_Time'] = df2['End_Time'].astype('datetime64[ns]') - df2['Start_Time'].astype('datetime64[ns]')

In [275]:
df2['Accident_Time']

0         0 days 05:14:00
1         0 days 00:30:00
2         0 days 00:30:00
3         0 days 00:30:00
4         0 days 00:30:00
                ...      
7728389   0 days 00:28:36
7728390   0 days 00:26:53
7728391   0 days 00:28:28
7728392   0 days 00:29:21
7728393   0 days 00:29:25
Name: Accident_Time, Length: 7728394, dtype: timedelta64[ns]

In [276]:
df2['End_Time'] = df2['End_Time'].astype('datetime64[ns]')
df2['Start_Time'] = df2['Start_Time'].astype('datetime64[ns]')

In [277]:
df2['End_Time']

0         2016-02-08 11:00:00
1         2016-02-08 06:37:59
2         2016-02-08 07:19:27
3         2016-02-08 07:53:34
4         2016-02-08 08:09:07
                  ...        
7728389   2019-08-23 18:32:01
7728390   2019-08-23 19:38:23
7728391   2019-08-23 19:28:49
7728392   2019-08-23 19:29:42
7728393   2019-08-23 19:21:31
Name: End_Time, Length: 7728394, dtype: datetime64[ns]

In [278]:
df['Start_Time']

0          2016-02-08 05:46:00
1          2016-02-08 06:07:59
2          2016-02-08 06:49:27
3          2016-02-08 07:23:34
4          2016-02-08 07:39:07
                  ...         
7728389    2019-08-23 18:03:25
7728390    2019-08-23 19:11:30
7728391    2019-08-23 19:00:21
7728392    2019-08-23 19:00:21
7728393    2019-08-23 18:52:06
Name: Start_Time, Length: 7728394, dtype: object

In [279]:
df.columns

Index(['ID', 'Source', 'Severity', 'Start_Time', 'End_Time', 'Start_Lat',
       'Start_Lng', 'End_Lat', 'End_Lng', 'Distance(mi)', 'Description',
       'Street', 'City', 'County', 'State', 'Zipcode', 'Country', 'Timezone',
       'Airport_Code', 'Weather_Timestamp', 'Temperature(F)', 'Wind_Chill(F)',
       'Humidity(%)', 'Pressure(in)', 'Visibility(mi)', 'Wind_Direction',
       'Wind_Speed(mph)', 'Precipitation(in)', 'Weather_Condition', 'Amenity',
       'Bump', 'Crossing', 'Give_Way', 'Junction', 'No_Exit', 'Railway',
       'Roundabout', 'Station', 'Stop', 'Traffic_Calming', 'Traffic_Signal',
       'Turning_Loop', 'Sunrise_Sunset', 'Civil_Twilight', 'Nautical_Twilight',
       'Astronomical_Twilight'],
      dtype='object')

---
### 5.3 Handling Accuracy & Validity


#### Temperature (F)
> Rounded to nearest integer and converted to `int32` for memory efficiency.


In [280]:
df2['Temperature(F)'] = round((df2['Temperature(F)']))

In [281]:
df2['Temperature(F)'].describe()

count    7.728394e+06
mean     6.171284e+01
std      1.881424e+01
min     -8.900000e+01
25%      5.000000e+01
50%      6.400000e+01
75%      7.600000e+01
max      2.070000e+02
Name: Temperature(F), dtype: float64

In [282]:
df2['Temperature(F)'] = df2['Temperature(F)'].astype('int32')

#### Pressure (in)
> Rounded to nearest integer and converted to `int16`.


In [283]:
df2['Pressure(in)']


0          30
1          30
2          30
3          30
4          30
           ..
7728389    29
7728390    29
7728391    30
7728392    30
7728393    29
Name: Pressure(in), Length: 7728394, dtype: int16

In [284]:
df2['Pressure(in)'] = round(df2['Pressure(in)'])

In [285]:
df2['Pressure(in)'] = df2['Pressure(in)'].astype('int16')

#### Visibility (mi)
> Rounded to nearest integer and converted to `int16`.


In [286]:
## Visisbility 

In [287]:
df2['Visibility(mi)'].describe()

count    7.728394e+06
mean     9.108135e+00
std      2.670951e+00
min      0.000000e+00
25%      1.000000e+01
50%      1.000000e+01
75%      1.000000e+01
max      1.400000e+02
Name: Visibility(mi), dtype: float64

In [288]:
df2['Visibility(mi)'] = round(df2['Visibility(mi)'])

In [289]:
df2['Visibility(mi)'] = df2['Visibility(mi)'].astype('int16')

#### Wind Direction — Full Name Mapping
> All abbreviated direction codes mapped to full names for clarity.


In [290]:
## wind direction

In [291]:
df2['Wind_Direction'].value_counts()

Wind_Direction
CALM                1505387
South                597364
West                 548664
North                460536
South South West     384840
East                 382459
West North West      378781
North West           369352
South West           364470
West South West      353806
South South East     349110
North North West     333427
South East           294901
East South East      268311
North East           258639
East North East      258474
North North East     255311
Varaible             250566
Name: count, dtype: int64

In [292]:
mapping = {
    'S' : 'South',
    'W' : 'West',
    'E' : 'East',
    'N' : 'North',
    'SSW' : 'South South West',
    'WNW' : 'West North West',
    'NW' :'North West',
    'SW' :'South West',
    'WSW' : 'West South West',
    'SSE': 'South South East',
    'NNW' : 'North North West',
    'SE' : 'South East',
    'ESE' : 'East South East',
    'NE' : 'North East',
    'ENE' : 'East North East',
    'NNE' : 'North North East',
    'CALM' : 'CALM',
    'VAR' : 'Varaible',
    'Varaible' : 'Varaible',
    'East' : 'East',
    'West' : "West",
    'South' : 'South',
    'North' : 'North'
}

In [293]:
df2['Wind_Direction']

0                      CALM
1                      CALM
2                South West
3                South West
4                South West
                 ...       
7728389                West
7728390          South West
7728391    South South West
7728392          South West
7728393          South West
Name: Wind_Direction, Length: 7728394, dtype: object

In [294]:
df2['Wind_Direction'] = df2['Wind_Direction'].map(mapping)

In [295]:
## Wind speed

In [296]:
df2['Wind_Speed(mph)'].describe()

count    7.728394e+06
mean     7.634823e+00
std      5.223725e+00
min      0.000000e+00
25%      5.000000e+00
50%      7.000000e+00
75%      1.000000e+01
max      1.087000e+03
Name: Wind_Speed(mph), dtype: float64

In [297]:
df2['Wind_Speed(mph)'] = df2['Wind_Speed(mph)'].clip(upper=300)

In [298]:
df2['Wind_Speed(mph)'].describe()

count    7.728394e+06
mean     7.633833e+00
std      5.119394e+00
min      0.000000e+00
25%      5.000000e+00
50%      7.000000e+00
75%      1.000000e+01
max      3.000000e+02
Name: Wind_Speed(mph), dtype: float64

In [299]:
df3 = df2.copy()

---
## ⚙️ 6. Feature Engineering

> Working on `df3` (copy of cleaned `df2`).  
> Goal: Create new meaningful columns that improve both analysis and ML model performance.

| Feature Group | New Columns | Purpose |
|---|---|---|
| **Time** | `hour`, `Day_Of_Week`, `Month_Name`, `Year`, `Season`, `is_weekend`, `time_of_day`, `is_Rush_Hour`, `Accident_Duration(min)` | Temporal patterns |
| **Severity** | `Severity` (label mapped) | Human-readable target |
| **Weather** | `Poor_Visibility`, `Is_Precipitating`, `Temp_Category` | Risk flags |
| **POI** | `Nearby_POI_Count`, `Has_Any_Poi` | Fix POI imbalance |

---


### 6.1 Time-Based Features

In [300]:
## Accident time
df3['Accident_Duration(min)'] = round((df3['End_Time'] - df3['Start_Time']).dt.total_seconds()/60)

In [301]:
## hour  
df3['hour'] = df3['Start_Time'].dt.hour

In [302]:
## day of week
df3['Day_Of_Week'] = df3['Start_Time'].dt.day_name()


In [303]:
## month
df3['Month_Name'] = df3['Start_Time'].dt.month_name()

In [304]:
## Year
df3['Year'] = df3['Start_Time'].dt.year

In [305]:
df3['Month_Number']=df3['Start_Time'].dt.month

In [306]:
## Season
season = {
    12:'Winter', 1:'Winter', 2:'Winter',
    3:'Spring',  4:'Spring', 5:'Spring',
    6:'Summer',  7:'Summer', 8:'Summer',
    9:'Fall',   10:'Fall',  11:'Fall'
}

In [307]:
df3['Season'] = df3['Month_Number'].map(season)

In [308]:
df3.drop(columns='Month_Number',inplace=True)

In [309]:
## is weekend
df3['is_weekend'] = df3['Start_Time'].dt.dayofweek >= 5

In [310]:
## Time of the day
df3['time_of_day'] =pd.cut(df3['hour'],
    bins=[0,6,12,17,21,24],
    labels=['Night','Morning','Afternoon','Evening','Late Night'])

In [311]:
## is rush hour
df3['is_Rush_Hour'] = df3['hour'].between(7,9) | df3['hour'].between(16,19)

### 6.2 Severity Label Mapping
> Mapped numeric severity (1–4) to descriptive labels: `Low`, `Medium`, `High`, `Critical`.


In [312]:
# ##is_sever
# df3['is_severe'] = df3['Severity'] >= 3

In [313]:
## severity
mapping = {
    1 : 'Low',
    2:'Medium',
    3 : 'High' ,
    4 : 'Critical'
}

In [314]:
## severity
df3['Severity'] = df3['Severity'].map(mapping)

### 6.3 Weather Risk Flags
> Binary flags to capture dangerous weather conditions clearly.


In [315]:
## Poor Visisbility
df3['Poor_Visibility'] = df3['Visibility(mi)'] < 1

In [316]:
## Is precipiting
df3['Is_Precipitaing'] = df3['Precipitation(in)'] > 0

In [317]:
## temprature
df3['Temp_Category'] = pd.cut(df['Temperature(F)'] , bins = [-50 , 32,60,85,150] , labels=['Freezing' , 'Cold' , 'Mild' , 'Hot'])

### 6.4 POI Aggregation
> All 12 POI boolean columns (heavily imbalanced individually) are combined into two meaningful features:
> - `Nearby_POI_Count` — total number of POIs near the accident
> - `Has_Any_Poi` — simple binary flag


In [318]:
## Nearby POI Count
cols = ['Amenity',
       'Bump', 'Crossing', 'Give_Way', 'Junction', 'No_Exit', 'Railway',
       'Roundabout', 'Station', 'Stop', 'Traffic_Calming', 'Traffic_Signal']

In [319]:
df3['Nearby_POI_Count'] = df[cols].sum(axis=1)

In [320]:
## has any poi
df3['Has_Any_Poi'] = df3['Nearby_POI_Count'] > 0

In [321]:
df3

,ID,Source,Severity,Start_Time,End_Time,Start_Lat,Start_Lng,End_Lat,End_Lng,Distance(mi),...,Year,Season,is_weekend,time_of_day,is_Rush_Hour,Poor_Visibility,Is_Precipitaing,Temp_Category,Nearby_POI_Count,Has_Any_Poi
0,1,Source2,High,2016-02-08 05:46:00,2016-02-08 11:00:00,39.865147,-84.058723,39.865147,-84.058723,0.010,...,2016,Winter,False,Night,False,False,True,Cold,0,False
1,2,Source2,Medium,2016-02-08 06:07:59,2016-02-08 06:37:59,39.928059,-82.831184,39.928059,-82.831184,0.010,...,2016,Winter,False,Night,False,False,False,Cold,0,False
2,3,Source2,Medium,2016-02-08 06:49:27,2016-02-08 07:19:27,39.063148,-84.032608,39.063148,-84.032608,0.010,...,2016,Winter,False,Night,False,False,False,Cold,1,True
3,4,Source2,High,2016-02-08 07:23:34,2016-02-08 07:53:34,39.747753,-84.205582,39.747753,-84.205582,0.010,...,2016,Winter,False,Morning,True,False,False,Cold,0,False
4,5,Source2,Medium,2016-02-08 07:39:07,2016-02-08 08:09:07,39.627781,-84.188354,39.627781,-84.188354,0.010,...,2016,Winter,False,Morning,True,False,False,Cold,1,True
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
7728389,7777757,Source1,Medium,2019-08-23 18:03:25,2019-08-23 18:32:01,34.002480,-117.379360,33.998880,-117.370940,0.543,...,2019,Summer,False,Evening,True,False,False,Hot,0,False
7728390,7777758,Source1,Medium,2019-08-23 19:11:30,2019-08-23 19:38:23,32.766960,-117.148060,32.765550,-117.153630,0.338,...,2019,Summer,False,Evening,True,False,False,Mild,0,False
7728391,7777759,Source1,Medium,2019-08-23 19:00:21,2019-08-23 19:28:49,33.775450,-117.847790,33.777400,-117.857270,0.561,...,2019,Summer,False,Evening,True,False,False,Mild,1,True
7728392,7777760,Source1,Medium,2019-08-23 19:00:21,2019-08-23 19:29:42,33.992460,-118.403020,33.983110,-118.395650,0.772,...,2019,Summer,False,Evening,True,False,False,Mild,0,False


---
## 💾 7. Saving Datasets

Two datasets saved:
- **`Cleaned_Dataset.csv`** — Full cleaned dataset with all original + new columns (for EDA & analysis)
- **`ML_Ready_Dataset.csv`** — Feature engineered version ready for ML modelling


In [322]:
df3.to_csv("Ml_Ready_Dataset.csv")

In [323]:
df2.to_csv("Clened_Dataset.csv")